### PR TITLE
Icache block size

### PR DESCRIPTION
**Pull request by tommydcjung**

### CODE REVIEW

File: v/vanilla_bean/vanilla_core.v [hunk 1]
```diff
@@ -1123,10 +1155,8 @@ module vanilla_core
     | stall_barrier;
 
   wire stall_all = stall_icache_store
```
Comment (dpetrisko) [inline v/vanilla_bean/vanilla_core.v:1157]: Are these removals related?

File: v/vanilla_bean/vanilla_core.v [hunk 2]
```diff
@@ -35,15 +38,15 @@ module vanilla_core
...
-    , pc_width_lp=(icache_tag_width_p+icache_addr_width_lp)
+    , pc_width_lp=(icache_tag_width_p+`BSG_SAFE_CLOG2(icache_entries_p))
     , reg_addr_width_lp = RV32_reg_addr_width_gp
     , data_mask_width_lp=(data_width_p>>3)
 
     , parameter debug_p=0
   )
   (
     input clk_i
+    , input network_reset_i
```
Comment (dpetrisko) [inline v/vanilla_bean/vanilla_core.v:49]: Why is a separate network_reset_i required?

File: v/vanilla_bean/icache.v [hunk 3]
```diff
@@ -174,10 +216,22 @@ module icache
   assign icache_flush_r_o = icache_flush_r;
 
 
+  // Energy-saving logic
+  // - Don't read the icache if the current pc is at the first word of the block, and 
+  //   there is a hint from the next-pc logic that it is reading pc+4 next (no branch or jump).
+  assign v_li = w_i
+    ? write_en_icache
+    : (v_i & ((&pc_r[0+:icache_block_offset_width_lp]) | ~read_pc_plus4_i));
```
Comment (dpetrisko) [inline v/vanilla_bean/icache.v:224]: If I'm reading correctly, this would only read if the PC is the last word of a line (0b11) or if non-consecutive. Should the comment be updated?

File: v/vanilla_bean/icache.v [hunk 4]
```diff
@@ -13,18 +13,21 @@ module icache
   import bsg_vanilla_pkg::*;
```
Comment (dpetrisko) [inline v/vanilla_bean/icache.v:13]: Is there a diagram somewhere I can look at?

File: v/vanilla_bean/vanilla_core.v (hunk 1 above)
Comment (tommydcjung) [reply v/vanilla_bean/vanilla_core.v:1159]: Just cleaning up the unnecessary logic before making big changes.

File: v/vanilla_bean/icache.v (hunk 4 above)
Comment (tommydcjung) [reply v/vanilla_bean/icache.v:15]: https://docs.google.com/presentation/d/1ZeRHYhqMHJQ0mRgDTilLuWQrZF7On-Be_KNNosgeW0c/edit#slide=id.g10d2e6febb9_1_0

File: v/vanilla_bean/vanilla_core.v (hunk 2 above)
Comment (tommydcjung) [reply v/vanilla_bean/vanilla_core.v:51]: Added comments

File: v/vanilla_bean/icache.v (hunk 3 above)
Comment (tommydcjung) [reply v/vanilla_bean/icache.v:226]: Updated.

File: v/vanilla_bean/vanilla_core.v (hunk 1 above)
Comment (dpetrisko) [reply v/vanilla_bean/vanilla_core.v:1159]: So these are hazards that no longer exist in the manycore, so we're removing the stall signals for them?

File: v/vanilla_bean/icache.v [hunk 5]
```diff
@@ -123,32 +122,77 @@ module icache
...
     ? branch_imm_val[RV32_Bimm_width_gp] 
     : jal_imm_val[RV32_Jimm_width_gp];
 
-  assign pc_lower_cout = write_branch_instr
+  wire pc_lower_cout = write_branch_instr
     ? branch_pc_lower_cout
     : jal_pc_lower_cout;
 
-  
+
```
Comment (dpetrisko) [inline v/vanilla_bean/icache.v:139]: Is this logic cheaper than using a byte-masked RAM?

File: v/bsg_manycore_pkg.v [hunk 6]
```diff
@@ -16,7 +16,7 @@ package bsg_manycore_pkg;
 
 
   localparam bsg_manycore_reg_id_width_gp = 5;
-
+  localparam bsg_manycore_icache_block_size_in_words_gp = 4;
```
Comment (dpetrisko) [inline v/bsg_manycore_pkg.v:19]: More of a 'sometime before merging' thing, but can we make this a parameter?

File: v/vanilla_bean/icache.v (hunk 5 above)
Comment (tommydcjung) [reply v/vanilla_bean/icache.v:139]: Yeah this is much cheaper

File: v/vanilla_bean/vanilla_core.v (hunk 1 above)
Comment (tommydcjung) [reply v/vanilla_bean/vanilla_core.v:1159]: you can go back and check, those stalls are always zero so the hazard never existed

File: v/vanilla_bean/icache.v (hunk 5 above)
Comment (dpetrisko) [reply v/vanilla_bean/icache.v:139]: nice

File: v/bsg_manycore_pkg.v (hunk 6 above)
Comment (tommydcjung) [reply v/bsg_manycore_pkg.v]: I made it a parameter at the top-level

Comment (dpetrisko) [reply v/bsg_manycore_pkg.v]: Great!

File: v/vanilla_bean/vanilla_core.v (hunk 1 above)
Comment (dpetrisko) [reply v/vanilla_bean/vanilla_core.v:1159]: kk

**Review by dpetrisko — APPROVED** [review]
